Interior penalty HDG Navier-Stokes improvements

lindsayad opened the pull request:
Closes #31399, https://github.com/idaholab/moose/issues/31955

Also adds
- channel MMS test
- backwards facing step example
- scripts for generating all results in our HDG NS paper
- bug fixes to make MOOSE HDG NS perform as well in preconditioning as with our firedrake implementation

lindsayad committed:
Fix MassFluxPenaltyIPHDG doc to include normals mult

Refs #31399

lindsayad committed:
Add _current_side_volume to HDGKernel

lindsayad committed:
Add missing precalculateResidual calls in ADIntegratedBC/ADDGKernel

Hopefully no one has overridden this method in downstream classes
because if so they've been ignored

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3176398/), step Generate coverage on 8a425de wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>417dcf</th>
      <th colspan=4>#31407 8a425d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.68%</td>
      <td>85.68%</td>
      <td>-0.00%</td>
      <td>85.71%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>118003</td>
      <td>118008</td>
      <td>+5</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19727</td>
      <td>19728</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31407/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31407/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>417dcf</th>
      <th colspan=4>#31407 8a425d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.49%</td>
      <td>85.51%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>19847</td>
      <td>19863</td>
      <td>+16</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3368</td>
      <td>3367</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31407/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31407/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31407/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.71% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Facet mass computations multiplied by h and jumps divided by h

Mass computation multiplied by h is consistent with the correct L2-norm on facets
according to Sander. With the correct norm we can get the correct scalings in
error analysis

lindsayad committed:
Work on bfs for ip-hdg

lindsayad committed:
Facet pressure should be same order as velocity

lindsayad committed:
Create new MassContinuity IP HDG classes

This changed our formulation slightly to match what is used in
our firedrake code. This does not change results, but it should
make it easier to construct simulations with Neumann boundary
conditions like the BFS

lindsayad committed:
Demonstrate MMS channel flow, e.g. Neumann BC

lindsayad committed:
Make prop_names and prop_values required in GenericVectorFunctorMaterial

lindsayad committed:
Make MassFluxPenaltyBC take functor instead of function argument

lindsayad committed:
Add backwards facing step test

Right now the Schur complement KSP convergence is poor

lindsayad committed:
upwind even on Dirichlet boundaries

lindsayad committed:
Use cell diameter instead of volume/area for h in diffusion

lindsayad committed:
Successfully matched Firedrake and libMesh/MOOSE

Changed to match shape functions. Use nodal Lagrange for the DG
primal variable and nodal Lagrange for the trace variable so this
actually becomes an EDG-HDG method

lindsayad committed:
No longer have unexplained superconvergence which I think is good

lindsayad committed:
Don't use method in tests spec

lindsayad committed:
Update framework IP HDG tests for recent changes

lindsayad committed:
Matching Firedrake

lindsayad committed:
Make alpha a required param so I don't forget it again

lindsayad committed:
Vector diffusion with Neumann boundary matching

lindsayad committed:
Stokes now matching

lindsayad committed:
Compute the advection term identically to firedrake

lindsayad committed:
Now matching with convection

lindsayad committed:
Convert MassFluxPenalty to use an assembly helper

This also enabled, which is likely a critical, adding the AL term
to the Neumann condition for the Lagrange multiplier trace variable

Also:
- Update NS inputs to new syntax
- Make `Assembly::sideElem()` return a reference to const pointer

lindsayad committed:
Now matching firedrake with AL

lindsayad committed:
BFS with our Schur complement strategy now converges like a boss

lindsayad committed:
Good bfs solver options with only outer ksp

lindsayad committed:
Modify bfs so we can ramp to a Reynolds number of 10k

lindsayad committed:
Add sequence time stepper for ramping up Reynolds

lindsayad committed:
Introduce HPDDM input file

lindsayad committed:
View ILU KSP

lindsayad committed:
Add script for three 2d bfs runs

lindsayad committed:
Work on building somewhat automated setup for paper result generation

- Split out hpddm and strumpack solver option inputs
  - This will enable reusing these options for the lid driven cavity
- Use lower triangular factorization for Schur in hpddm and strumpack
- Write out mesh used in outputs (log, profile, MOOSE Outputs)
- Change to using dedicated KSPs for solving A and S
  - This allows us to diagnose what components of the solve are working
    well and what are not
- Check-in hpddm run script
- Use 150 singular vectors per subdomain per hpddm
  - This allows the Re = 1 solve to converge much better for the twice
    refined coarse06.msh

Uptick to 150 singular vectors per subdomain

lindsayad committed:
Get lid driven input for paper results running

lindsayad committed:
Get 3d lid driven cavity running

Fascinatingly jacobi rowsum does not work at all with the mass matrix
for preconditioning the Schur complement, but the diagonal seems fine

lindsayad committed:
Create 3d ldc run script

lindsayad committed:
Prepare script for running 2D HDG ldc cases

lindsayad committed:
Get some timing and profile data for the 2D LDC runs

Things seem to be running very slowly, so I want some diagnostic
data

lindsayad committed:
Python script for parsing iteration count data from log files

lindsayad committed:
Use coarse12 mesh in bfs 2d

lindsayad committed:
Switch 2d bfs to be HDG

lindsayad committed:
Add more functionality to python script

Add output for num elems, num dofs, num condensed dofs, and average
solve time

lindsayad committed:
Update python script to handle num elems for multiprocess logs

lindsayad committed:
Handle ksp solver divergence

lindsayad committed:
Get script output condensed dof counts correct

lindsayad committed:
Add ability to create Schur pre from superposition of A11 and M

lindsayad committed:
Convert run script to python

lindsayad committed:
Only add num_steps to moose command line if specified

lindsayad committed:
Allow running detached jobs

lindsayad committed:
Include schur pre as part of tag

lindsayad committed:
Prepare bfs 2d case for a11_and_mass pre run

lindsayad committed:
Not including 5000 Re results at the moment

lindsayad committed:
Get 3d bfs test case running

lindsayad committed:
Destroy and re-create pressure mass matrix every iteration when adding A11

lindsayad committed:
Need complete order for 3d bfs to store side hierarchic dofs

lindsayad committed:
EDGHDG must still use discontinuous facets for pressure

lindsayad committed:
Match firedrake nonlinear solver settings

lindsayad committed:
n=6 leads to failing nonlinear solve for 3d lid driven cavity

lindsayad committed:
Must use tet14 for 3d ldc in order to store side hierarchic dofs

lindsayad committed:
Try using asm because global strumpack ilu is really failing on 3d problems

lindsayad committed:
Don't require proc multiplier cli arg if only running a single case

lindsayad committed:
Add run.py script for 3d ldc

lindsayad committed:
Try n = 8

lindsayad committed:
Do EDG-HDG 2D testing

lindsayad committed:
Only allow 30 KSP outer iterations

lindsayad committed:
Back to HDG

lindsayad committed:
Switch to KSP(hat(S))

lindsayad committed:
Add discretization argument to run python script

This removes the need for git diffs when toggling back and forth
between HDG and EDG-HDG.

Also remove the detach parser argument

lindsayad committed:
Add disc to log file name

lindsayad committed:
Omit hat(S) iteration counts

lindsayad committed:
Output normalized solve average

lindsayad committed:
Remove files unused for paper

lindsayad committed:
test paper cases

lindsayad committed:
Fix unicode characters

lindsayad committed:
Fix HDG moose test inputs

lindsayad committed:
Fix navier_stokes IP-H tests

- Do more refinements in MMS testing to get asymptotic value for pressure
- bump alpha for proper stability

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3421910/), step Versioner verify on 4b617f6 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 10 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| petsc     | CHANGE   | conda/petsc/meta.yaml                   |
| petsc     | CHANGE   | petsc                                   |
| libmesh   | CHANGE   | conda/libmesh/conda_build_config.yaml   |
| libmesh   | CHANGE   | conda/libmesh/meta.yaml                 |
| libmesh   | CHANGE   | libmesh                                 |
| moose-dev | CHANGE   | apptainer/moose-dev.def                 |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |
| moose-dev | CHANGE   | framework/contrib/mfem                  |
| moose-dev | CHANGE   | scripts/update_and_rebuild_mfem.sh      |

### Versioner versions

Found 9 packages, 3 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version                 |
|-----------|----------|---------|-----------|--------------------|----------------------------|
| petsc     | CHANGE   | 09c8ce6 | 09499e8   | 3.24.0 build 1     | 3.24.2 build 0             |
| libmesh   | CHANGE   | e987a26 | f6e51eb   | 2025.10.15 build 1 | 2025.11.25_c144a6e build 0 |
| moose-dev | CHANGE   | a20764f | 8b8bd90   | 2025.11.24         | 2025.11.26                 |

lindsayad committed:
In matrix testing, verify MassMatrixHDG instead of DGKernel

This is important because MassMatrixHDG is what is used in production
runs

lindsayad committed:
Demonstrate proper setting of mass matrix density in AL setting

lindsayad committed:
Do proper scaling of -1 for Schur complement in matrix testing

This ensures that -M/gamma is indeed spectrally equivalent
to our Schur complement as we expect

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3422041/), step Results summary on eb95876 wanted to post the following:

## Framework test summary

Compared against 82bb60c in job [civet.inl.gov/job/3419953](https://civet.inl.gov/job/3419953).

No change
## Modules test summary

Compared against 82bb60c in job [civet.inl.gov/job/3419953](https://civet.inl.gov/job/3419953).

### Added tests

| Test                                                                           | Time (s)   |
|--------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_element/ins/hdg/ip/paper-cases/bfs.csv`             | 23.57      |
| `navier_stokes/test:finite_element/ins/hdg/ip/paper-cases/lid-driven.csv`      | 16.22      |
| `navier_stokes/test:finite_element/ins/hdg/ip/channel-flow.channel-mms`        | SKIP       |
| `navier_stokes/test:finite_element/ins/hdg/ip/channel-flow.channel-mms-second` | SKIP       |

lindsayad committed:
Remove extraneous firedrake matching input

lindsayad committed:
Add missing doc pages

Also fix registration of objects

lindsayad committed:
Fix references to use_pressure_mass_matrix

lindsayad committed:
Remember to set max_threads for SC

lindsayad committed:
Allow restep testing of paper cases

lindsayad committed:
Add IP-HDG navier stokes doc

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 910736a3a20aded528c3ba99371ce65980dc8e2c
+Subproject commit c144a6e3ae9d721c0b780ac0035cff984c742ea1
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3428320/), step Results summary on 17b3a38 wanted to post the following:

## Framework test summary

Compared against d67480e in job [civet.inl.gov/job/3424342](https://civet.inl.gov/job/3424342).

No change
## Modules test summary

Compared against d67480e in job [civet.inl.gov/job/3424342](https://civet.inl.gov/job/3424342).

### Added tests

| Test                                                                           | Time (s)   |
|--------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_element/ins/hdg/ip/paper-cases/bfs.csv`             | 22.83      |
| `navier_stokes/test:finite_element/ins/hdg/ip/paper-cases/lid-driven.csv`      | 15.77      |
| `navier_stokes/test:finite_element/ins/hdg/ip/channel-flow.channel-mms-second` | SKIP       |
| `navier_stokes/test:finite_element/ins/hdg/ip/channel-flow.channel-mms`        | SKIP       |

moosebuild commented:
Job [Clang min](https://civet.inl.gov/job/3428283/) on 17b3a38 : invalidated by @lindsayad

lindsayad commented:
Disabling testing until https://github.com/idaholab/moose/pull/31906 goes in

lindsayad commented:
It would be nice to get a review on this at some point

GiudGiud commented:
Will do after the training for sure

GiudGiud commented on framework/include/hdgkernels/MassContinuityAssemblyHelper.h:
```diff
@@ -0,0 +1,59 @@
+  virtual void lmFace() override;
+  virtual void scalarDirichlet(const Moose::Functor<Real> &) override
+  {
+    mooseError("We do not assign Dirichlet values for pressure");
```
maybe reword? an outlet BC at fixed pressure is a fairly common use

GiudGiud commented on framework/include/hdgkernels/MassContinuityAssemblyHelper.h:
```diff
@@ -0,0 +1,59 @@
+  /// The radial coordinate index for RZ coordinate systems
+  const unsigned int _rz_radial_coord;
+
+  /// The velocities on the element interior
```
```suggestion
  /// The velocities at interior element Qps on the element interior
```
if that s true

GiudGiud commented on framework/src/bcs/MassContinuityIPHDGBC.C:
```diff
@@ -0,0 +1,37 @@
+#include "MassContinuityIPHDGBC.h"
+#include "MassContinuityAssemblyHelper.h"
+
+registerMooseObject("NavierStokesApp", MassContinuityIPHDGBC);
```
```suggestion
registerMooseObject("MooseApp", MassContinuityIPHDGBC);
```
should have complained at doco build / SQA check time :O

GiudGiud reviewed (commented):
I ll review more later, got to head back to boston

lindsayad commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/paper-cases/bfs/bfs-2d-root.i:
```diff
@@ -0,0 +1,430 @@
+    type = MassMatrixHDG
+    variable = pressure_bar
+    matrix_tags = 'mass'
+    density = '${fparse -1/gamma}'
```
$\gamma$ is discussed in `MassFluxPenaltyIPHDG.md`

lindsayad committed:
Address Guillaume review

- Better doxygen and commenting
- Move Mass IPHDG objects to Navier-Stokes module
- Remove largely redundant second order MMS channel test case
- Rename strumpack solver options input to fs-plus-strumpack-preconditioner

lindsayad commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/paper-cases/fs-plus-strumpack-preconditioner.i:
```diff
@@ -0,0 +1,25 @@
+[Preconditioning]
```
renamed

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/paper-cases/parse-log.py:
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
+
+# Courtesy chatgpt-5
```
> Just removing something because it was created through (iterations with) an LLM doesn't resonate with me.

resonates with me. "ChatGPT can write a MOOSE BodyForce kernel if you ask it to, so we didn't bother to check one in" is exactly the argument.

Can we "RunApp" (RunCommand really) this script to get some form of testing?

lindsayad replied:
> > Just removing something because it was created through (iterations with) an LLM doesn't resonate with me.
> 
> resonates with me. "ChatGPT can write a MOOSE BodyForce kernel if you ask it to, so we didn't bother to check one in" is exactly the argument.

I don't understand what you're saying. You're saying that we shouldn't have `BodyForce` in the repo?

GiudGiud replied:
if the tool can re-generate it on the fly (which we are not quite there, but maybe not that far with an AI-vscode plugin) we would not need it in the repo

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3500808/), step Results summary on 7cbf80d wanted to post the following:

## Framework test summary

Compared against 5eee4e5 in job [civet.inl.gov/job/3497366](https://civet.inl.gov/job/3497366).

No change
## Modules test summary

Compared against 5eee4e5 in job [civet.inl.gov/job/3497366](https://civet.inl.gov/job/3497366).

### Added tests

| Test                                                                           | Time (s)   | Memory (MB)   |
|--------------------------------------------------------------------------------|------------|---------------|
| `navier_stokes/test:finite_element/ins/hdg/ip/paper-cases/bfs.csv`             | 22.31      | 510.54        |
| `navier_stokes/test:finite_element/ins/hdg/ip/paper-cases/lid-driven.csv`      | 15.37      | 368.43        |
| `navier_stokes/test:finite_element/ins/hdg/ip/channel-flow.channel-mms-second` | SKIP       | 0.00          |
| `navier_stokes/test:finite_element/ins/hdg/ip/channel-flow.channel-mms`        | SKIP       | 0.00          |

GiudGiud commented on modules/navier_stokes/doc/content/source/bcs/NavierStokesStressIPHDGPrescribedFluxBC.md:
```diff
@@ -10,6 +10,9 @@ and
 !listing modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/channel-flow/mms-channel.i block=BCs/momentum_y_diffusion_neumann
 
+When applying a zero traction condition at an outflow, if the flow is fully developed, then
+implicitly this applies a zero pressure condition.
```
yes no mass flow

GiudGiud replied:
seems like it's not this if it's an outflow condition though

lindsayad committed:
NavierStokesStressIPHDGPrescribedFluxBC -> NavierStokesStressIPHDGPrescribedTractionBC

lindsayad committed:
Remove python files

lindsayad committed:
test -> trial for describing variable shapes

GiudGiud reviewed (approved):
pending issue creation for functor block restriction check (maybe we already have one btw)

lindsayad commented:
Module debug failures are unrelated
```
[299.9s] [ 182MB]  TIMEOUT stochastic_tools/test:web_server_control.stochastic_control/batch_restore FAILED (TIMEOUT) [stochastic_tools-dbg]
[1.054s] [  69MB]     FAIL porous_flow/test:newton_cooling.nc06 FAILED (CRASH) [porous_flow-dbg]
```
CUDA GPU failures also unrelated.

Thanks for your review @GiudGiud!
